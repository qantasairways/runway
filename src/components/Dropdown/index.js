/* eslint-disable jsx-a11y/label-has-for, jsx-a11y/label-has-associated-control */
import * as React from 'react';
import PropTypes from 'prop-types';
import Downshift from 'downshift';
import { css } from 'emotion';
import { KEY_CODE_SPACE, KEY_CODE_ENTER } from '../../constants/keyCodes';

import Menu from './components/Menu';
import MenuItem from './components/MenuItem';
import SelectOnKeyPressContainer from './components/SelectOnKeyPressContainer';
import TickIcon from '../../icons/Tick';
import ChevronDown from '../../icons/ChevronDown';
import noop from '../../utils/noop';
import { colours, layout, fontWeight } from '../../theme/airways';

export function dropdownStyles(
  { highlighted, growMenu, inline, height },
  isFocused,
  isOpen
) {
  return css({
    label: 'runway-dropdown',
    fontFamily: 'Ciutadella',
    fontSize: '18px',
    fontWeight: highlighted ? fontWeight.bold : fontWeight.regular,
    fontStyle: 'normal',
    fontStretch: 'normal',
    lineHeight: 1.56,
    letterSpacing: 'normal',
    background: colours.mediumGrey,
    borderColor: colours.darkeGrey,
    color: highlighted ? colours.highlights : colours.white,
    textDecoration: highlighted ? 'underline' : 'none',
    position: growMenu ? 'static' : 'relative',
    width: inline ? 'fit-content' : '100%',
    height,
    maxWidth: '100%',
    outline: isFocused && !isOpen ? '-webkit-focus-ring-color auto 5px' : 'none'
  });
}

export function inputWrapperStyles() {
  return css({
    label: 'runway-dropdown__input-wrapper',
    cursor: 'pointer',
    display: 'flex',
    width: '100%',
    height: '100%',
    alignItems: 'center',
    boxSizing: 'border-box',
    justifyContent: 'space-between'
  });
}

export function inputStyles({ inline }) {
  return css({
    border: '0px',
    clip: 'rect(0px, 0px, 0px, 0px)',
    margin: '-1px',
    position: 'absolute',
    label: 'runway-dropdown__input',
    backgroundColor: 'transparent',
    color: inline ? 'transparent' : 'inherit',
    cursor: 'pointer',
    width: '100%',
    height: '100%',
    '::placeholder': {
      color: 'inherit'
    }
  });
}

export function inputSvgStyles({ inline, highlighted }) {
  return css({
    label: 'runway-dropdown__input-svg',
    width: '24px',
    height: '100%',
    fill: highlighted ? colours.highlights : colours.white,
    verticalAlign: 'middle',
    padding: inline ? 0 : `0 ${layout.gutter} 0 0`,
    boxSizing: 'content-box'
  });
}

export function itemSvgContainerStyles({ selected }) {
  const visibility = selected ? 'visible' : 'hidden';

  return css({
    label: 'runway-dropdown__item-svg-container',
    visibility,
    marginRight: '10px'
  });
}

export function itemContainerStyles() {
  return css({
    label: 'runway-dropdown__item-container',
    display: 'inline-flex'
  });
}

export function itemSvgStyles() {
  return css({
    label: 'runway-dropdown__item-svg',
    width: '24px',
    fill: colours.darkerGrey
  });
}

function defaultItemToString(item) {
  return item === null ? '' : String(item);
}

function resolveItemAsString(itemToString = defaultItemToString, item) {
  return itemToString(item);
}

function generateMenu(menuProps) {
  const {
    isOpen,
    items,
    getItemProps,
    selectedItem,
    renderItem,
    itemToString,
    highlightedIndex,
    getMenuProps,
    focus,
    width
  } = menuProps;

  const list = items.map((item, index) => {
    const itemString = resolveItemAsString(itemToString, item);

    const props = {
      highlighted: highlightedIndex === index,
      selected: selectedItem.name === item.name,
      key: `${itemString}-${index}`,
      item
    };

    const itemProps = getItemProps({ ...props });

    return (
      <MenuItem {...itemProps}>{renderItem(item, index, itemProps)}</MenuItem>
    );
  });

  return (
    <Menu {...getMenuProps({ focus, width })} isOpen={isOpen}>
      {list}
    </Menu>
  );
}

function Render(props) {
  const {
    items,
    renderItem,
    isFocused,
    setFocus,
    label,
    placeholder,
    downshiftProps,
    menuWidth,
    inline
  } = props;

  const {
    isOpen,
    getItemProps,
    selectedItem,
    highlightedIndex,
    openMenu,
    getInputProps,
    getLabelProps,
    getMenuProps,
    closeMenu,
    itemToString
  } = downshiftProps;

  const menuProps = {
    isOpen,
    items,
    getItemProps,
    selectedItem,
    renderItem,
    itemToString,
    highlightedIndex,
    getMenuProps,
    focus: isFocused,
    width: menuWidth
  };

  const inputProps = getInputProps();

  const containerProps = {
    onClick: () => {
      openMenu();
    }
  };

  return (
    <div css={{ width: '100%', height: '100%' }}>
      <label {...getLabelProps()}>{label}</label>
      <div {...containerProps} css={inputWrapperStyles(props)}>
        <span
          css={{
            overflow: 'hidden',
            textOverflow: 'ellipsis',
            whiteSpace: 'nowrap',
            padding: inline ? 0 : `0 0 0 ${layout.gutter}`
          }}
        >
          {inputProps.value}
        </span>
        <input
          css={inputStyles(props)}
          readOnly
          tabIndex="0"
          placeholder={placeholder}
          {...getInputProps({
            onBlur: () => {
              closeMenu();
              setFocus(false);
            },
            onKeyDown: event => {
              const { keyCode } = event;
              if (keyCode === KEY_CODE_SPACE || keyCode === KEY_CODE_ENTER) {
                event.preventDefault();
                event.stopPropagation();
                setFocus(false);
                openMenu();
              }
            },
            onFocus: () => {
              setFocus(true);
            }
          })}
        />
        <span css={inputSvgStyles(props)}>
          <ChevronDown css={inputSvgStyles(props)} />
        </span>
      </div>
      {generateMenu(menuProps)}
    </div>
  );
}

Render.propTypes = {
  items: PropTypes.arrayOf(PropTypes.object),
  renderItem: PropTypes.func,
  focus: PropTypes.bool,
  label: PropTypes.string,
  placeholder: PropTypes.string,
  menuWidth: PropTypes.string,
  height: PropTypes.string,
  highlighted: PropTypes.bool,
  inline: PropTypes.bool,
  growMenu: PropTypes.bool,
  downshiftProps: PropTypes.shape({
    isOpen: PropTypes.bool,
    getItemProps: PropTypes.func,
    selectedItem: PropTypes.object,
    highlightedIndex: PropTypes.number,
    openMenu: PropTypes.func,
    getInputProps: PropTypes.func,
    getLabelProps: PropTypes.func,
    getMenuProps: PropTypes.func,
    closeMenu: PropTypes.func,
    itemToString: PropTypes.func
  }).isRequired
};

Render.defaultProps = {
  items: [],
  renderItem: noop,
  focus: false,
  label: '',
  placeholder: '',
  menuWidth: null,
  height: null,
  highlighted: false,
  inline: false,
  growMenu: false
};

export default class Dropdown extends React.Component {
  state = {
    isFocused: false
  };

  // eslint-disable-next-line consistent-return
  componentDidUpdate = prevProps => {
    if (this.itemsChanged(prevProps.items, this.props.items)) {
      if (
        this.validItem(this.downshiftSelectedItem) &&
        this.itemsContains(this.downshiftSelectedItem)
      )
        return this.downshiftSelectItem(this.downshiftSelectedItem);
      if (
        this.validItem(this.props.defaultItemWhenNoneSelected) &&
        this.itemsContains(this.props.defaultItemWhenNoneSelected)
      )
        return this.downshiftSelectItem(this.props.defaultItemWhenNoneSelected);
      return this.downshiftSelectItem(this.props.items[0]);
    }
  };

  setFocus = isFocused => {
    this.setState({ isFocused });
  };

  itemsContains = target =>
    !!this.props.items.filter(item => item.value === target.value).length;

  itemsChanged = (prevItems, nextItems) =>
    prevItems.length !== nextItems.length ||
    !prevItems.every(prevItem =>
      nextItems.some(nextItem => nextItem.value === prevItem.value)
    );

  collectSelectItem = selectItem => {
    this.downshiftSelectItem = selectItem;
  };

  collectSelectedItem = selectedItem => {
    this.downshiftSelectedItem = selectedItem;
  };

  getInitialSelectedItem = () =>
    this.validItem(this.props.initialSelectedItem)
      ? this.props.initialSelectedItem
      : this.props.items[0];

  validItem = item =>
    item && typeof item.name === 'string' && typeof item.value === 'string';

  render() {
    const { isFocused } = this.state;
    const { props, setFocus } = this;
    const { isOpen } = this;

    return (
      <Downshift
        {...props.downShiftProps}
        onChange={(selectedItem, stateAndHelpers) => {
          if (typeof props.onChange === 'function') {
            props.onChange(selectedItem, stateAndHelpers);
          }
        }}
        initialSelectedItem={this.getInitialSelectedItem()}
      >
        {downshiftProps => {
          const renderProps = {
            ...props,
            isOpen,
            isFocused,
            setFocus,
            downshiftProps
          };

          this.collectSelectItem(renderProps.downshiftProps.selectItem);
          this.collectSelectedItem(renderProps.downshiftProps.selectedItem);
          return (
            <div css={dropdownStyles(props, isFocused, isOpen)}>
              <SelectOnKeyPressContainer
                items={props.items}
                itemToString={props.downShiftProps.itemToString}
                focus={isFocused}
                downshiftProps={downshiftProps}
              >
                <Render {...renderProps} />
              </SelectOnKeyPressContainer>
            </div>
          );
        }}
      </Downshift>
    );
  }
}

function renderDefaultItem(item, index, props) {
  return (
    <span css={itemContainerStyles(props)}>
      <span css={itemSvgContainerStyles(props)}>
        <TickIcon css={itemSvgStyles(props)} />
      </span>
      <span css={{ whiteSpace: 'normal', wordWrap: 'break-word' }}>
        {item.name}
      </span>
    </span>
  );
}

Dropdown.propTypes = {
  /** Array of list items for the dropdown */
  items: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired
    })
  ).isRequired,
  /** Prop to render each list item. Must return an element. */
  renderItem: PropTypes.func,
  /** @ignore */
  focus: PropTypes.bool,
  /** @ignore */
  downShiftProps: PropTypes.shape({
    itemToString: PropTypes.func
  }),
  /** Optional string to set the width of the menu */
  menuWidth: PropTypes.string,
  /** Optional string to set the height of the dropdown */
  height: PropTypes.string,
  /** Triggered when the user changes the value
   *
   * @param {Object} selectedItem New value
   * @param {Object} stateAndHelpers stateAndHelpers object from Downshift */
  onChange: PropTypes.func,
  /** Optional flag to use highlighted styles */
  highlighted: PropTypes.bool,
  /** Optional flag to make the dropdown inline rather than full width */
  inline: PropTypes.bool,
  /**
   * Optional flag to allow the dropdown menu to grow larger than the dropdown container
   * Setting this flag to true requires a parent to have `position:absolute;` applied */
  growMenu: PropTypes.bool
};

Dropdown.defaultProps = {
  renderItem: renderDefaultItem,
  focus: false,
  downShiftProps: {
    itemToString: selectedItem => {
      if (!selectedItem || !selectedItem.name) {
        return '';
      }
      return String(selectedItem.name);
    }
  },
  menuWidth: null,
  height: '100%',
  onChange: null,
  highlighted: false,
  inline: false,
  growMenu: false
};
