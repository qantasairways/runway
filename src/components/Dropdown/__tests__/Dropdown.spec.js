import React from 'react';
import { mount } from 'enzyme';

import Dropdown from '../index';

jest.mock('shortid', () => ({ generate: () => 'mockId' }));

describe('Dropdown', () => {
  let component;

  const apple = { name: 'name:apple', value: 'value:apple' };
  const orange = { name: 'name:orange', value: 'name:orange' };
  const banana = { name: 'name:banana', value: 'name:banana' };
  const pear = { name: 'name:pear', value: 'value:pear' };

  const items = [apple, orange, banana];

  const props = {
    items,
    renderItem: () => <div>mockRenderItemResult</div>,
    withPadding: true,
    focus: true,
    downShiftProps: { itemToString: item => item.name },
    menuWidth: '800px',
    height: '20px',
    onChange: () => {},
    highlighted: true,
    inline: true,
    growMenu: true
  };

  it('renders correctly with defaults', () => {
    component = mount(<Dropdown items={[]} />);
    expect(component).toMatchSnapshot();
  });

  it('renders correctly with props provided', () => {
    component = mount(<Dropdown {...props} />);
    expect(component).toMatchSnapshot();
  });

  it('initially renders with valid initialSelectedItem selected', () => {
    component = mount(<Dropdown {...props} initialSelectedItem={orange} />);
    const input = component.find('input');
    expect(input.props().value).toEqual(orange.name);
  });

  it('initially renders with first available item selected when initialSelectedItem not valid', () => {
    const missingName = { value: 'value:bad' };
    const mockItems = [...items, missingName];
    component = mount(
      <Dropdown
        {...props}
        items={mockItems}
        initialSelectedItem={missingName}
      />
    );
    const input = component.find('input');
    expect(input.props().value).toEqual(items[0].name);
  });

  describe('when items change', () => {
    describe('when current selected is valid AND still exists in items', () => {
      it('should keep current selected', () => {
        component = mount(
          <Dropdown {...props} items={items} initialSelectedItem={orange} />
        );
        const nextItems = [pear, orange];
        component.setProps({ items: nextItems });
        component.update();
        const input = component.find('input');
        expect(input.props().value).toEqual(orange.name);
      });
    });
    describe('when current selected is does NOT exist in items AND defaultItemWhenNoneSelected is valid AND still exists in items', () => {
      it('should keep select defaultItemWhenNoneSelected', () => {
        component = mount(
          <Dropdown
            {...props}
            items={items}
            initialSelectedItem={orange}
            defaultItemWhenNoneSelected={banana}
          />
        );
        const nextItems = [apple, banana];
        component.setProps({ items: nextItems });
        component.update();
        const input = component.find('input');
        expect(input.props().value).toEqual(banana.name);
      });
    });
    describe('when current selected does NOT exist in items AND defaultItemWhenNoneSelected does NOT exist in items', () => {
      it('should select first available item', () => {
        component = mount(
          <Dropdown
            {...props}
            items={items}
            initialSelectedItem={orange}
            defaultItemWhenNoneSelected={banana}
          />
        );
        const nextItems = [apple, pear];
        component.setProps({ items: nextItems });
        component.update();
        const input = component.find('input');
        expect(input.props().value).toEqual(apple.name);
      });
    });
  });
});
