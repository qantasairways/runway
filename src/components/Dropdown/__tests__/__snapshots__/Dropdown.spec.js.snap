// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown renders correctly with defaults 1`] = `
.emotion-8 {
  font-family: Ciutadella;
  font-size: 18px;
  font-weight: 400;
  font-style: normal;
  font-stretch: normal;
  line-height: 1.56;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
  background: #3c3c3c;
  color: #fff;
  -webkit-text-decoration: none;
  text-decoration: none;
  position: relative;
  width: 100%;
  height: 100%;
  max-width: 100%;
  outline: none;
}

.emotion-7 {
  width: 100%;
  height: 100%;
}

.emotion-5 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  height: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.emotion-0 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  padding: 0 0 0 15px;
}

.emotion-1 {
  border: 0px;
  -webkit-clip: rect(0px,0px,0px,0px);
  clip: rect(0px,0px,0px,0px);
  margin: -1px;
  position: absolute;
  background-color: transparent;
  color: inherit;
  cursor: pointer;
  width: 100%;
  height: 100%;
}

.emotion-1::-webkit-input-placeholder {
  color: inherit;
}

.emotion-1::-moz-placeholder {
  color: inherit;
}

.emotion-1:-ms-input-placeholder {
  color: inherit;
}

.emotion-1::placeholder {
  color: inherit;
}

.emotion-2 {
  width: 24px;
  height: 100%;
  fill: #fff;
  vertical-align: middle;
  padding: 0 15px 0 0;
  box-sizing: content-box;
}

.emotion-6 {
  display: none;
  min-width: 240px;
  border-radius: 4px;
  box-shadow: 0 1px 2px 0 rgba(0,0,0,0.2);
  border: solid 1px #dadada;
  background-color: #ffffff;
  list-style: none;
  position: absolute;
  z-index: 5;
  margin: 0;
  top: 0px;
  left: 0px;
  width: 100%;
  box-sizing: border-box;
  padding: 10px;
}

<Dropdown
  downShiftProps={
    Object {
      "itemToString": [Function],
    }
  }
  focus={false}
  growMenu={false}
  height="100%"
  highlighted={false}
  inline={false}
  items={Array []}
  menuWidth={null}
  onChange={null}
  renderItem={[Function]}
>
  <Downshift
    defaultHighlightedIndex={null}
    defaultIsOpen={false}
    environment={[Window]}
    getA11yStatusMessage={[Function]}
    itemToString={[Function]}
    onChange={[Function]}
    onInputValueChange={[Function]}
    onOuterClick={[Function]}
    onSelect={[Function]}
    onStateChange={[Function]}
    onUserAction={[Function]}
    scrollIntoView={[Function]}
    selectedItemChanged={[Function]}
    stateReducer={[Function]}
    suppressRefError={false}
  >
    <div
      aria-expanded={false}
      aria-haspopup="listbox"
      aria-labelledby="downshift-0-label"
      aria-owns={null}
      className="emotion-8"
      role="combobox"
    >
      <SelectOnKeyPressContainer
        downshiftProps={
          Object {
            "clearItems": [Function],
            "clearSelection": [Function],
            "closeMenu": [Function],
            "getInputProps": [Function],
            "getItemProps": [Function],
            "getLabelProps": [Function],
            "getMenuProps": [Function],
            "getRootProps": [Function],
            "getToggleButtonProps": [Function],
            "highlightedIndex": null,
            "id": "downshift-0",
            "inputValue": "",
            "isOpen": false,
            "itemToString": [Function],
            "openMenu": [Function],
            "reset": [Function],
            "selectHighlightedItem": [Function],
            "selectItem": [Function],
            "selectItemAtIndex": [Function],
            "selectedItem": null,
            "setHighlightedIndex": [Function],
            "setItemCount": [Function],
            "setState": [Function],
            "toggleMenu": [Function],
            "unsetItemCount": [Function],
          }
        }
        focus={false}
        itemToString={[Function]}
        items={Array []}
      >
        <Render
          downShiftProps={
            Object {
              "itemToString": [Function],
            }
          }
          downshiftProps={
            Object {
              "clearItems": [Function],
              "clearSelection": [Function],
              "closeMenu": [Function],
              "getInputProps": [Function],
              "getItemProps": [Function],
              "getLabelProps": [Function],
              "getMenuProps": [Function],
              "getRootProps": [Function],
              "getToggleButtonProps": [Function],
              "highlightedIndex": null,
              "id": "downshift-0",
              "inputValue": "",
              "isOpen": false,
              "itemToString": [Function],
              "openMenu": [Function],
              "reset": [Function],
              "selectHighlightedItem": [Function],
              "selectItem": [Function],
              "selectItemAtIndex": [Function],
              "selectedItem": null,
              "setHighlightedIndex": [Function],
              "setItemCount": [Function],
              "setState": [Function],
              "toggleMenu": [Function],
              "unsetItemCount": [Function],
            }
          }
          focus={false}
          growMenu={false}
          height="100%"
          highlighted={false}
          inline={false}
          isFocused={false}
          items={Array []}
          label=""
          menuWidth={null}
          onChange={null}
          placeholder=""
          renderItem={[Function]}
          setFocus={[Function]}
        >
          <div
            className="emotion-7"
          >
            <label
              htmlFor="downshift-0-input"
              id="downshift-0-label"
            />
            <div
              className="emotion-5"
              onClick={[Function]}
            >
              <span
                className="emotion-0"
              />
              <input
                aria-activedescendant={null}
                aria-autocomplete="list"
                aria-controls={null}
                aria-labelledby="downshift-0-label"
                autoComplete="off"
                className="emotion-1"
                id="downshift-0-input"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                placeholder=""
                readOnly={true}
                tabIndex="0"
                value=""
              />
              <span
                className="emotion-2"
              >
                <ChevronDown
                  className="emotion-2"
                  height="24"
                  width="24"
                >
                  <svg
                    className="emotion-2"
                    height="24"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                    xmlnsXlink="http://www.w3.org/1999/xlink"
                  >
                    <defs>
                      <path
                        d="M2.12 1.29L6 5.17l3.88-3.88a.996.996 0 1 1 1.41 1.41L6.7 7.29a.996.996 0 0 1-1.41 0L.7 2.7a.996.996 0 0 1 0-1.41c.39-.38 1.03-.39 1.42 0z"
                        id="mockId"
                      />
                    </defs>
                    <g
                      fillRule="evenodd"
                      transform="translate(6 8)"
                    >
                      <mask
                        id="mockId"
                      >
                        <use
                          xlinkHref="#mockId"
                        />
                      </mask>
                      <use
                        fillRule="nonzero"
                        xlinkHref="#mockId"
                      />
                      <g
                        mask="url(#mockId)"
                      >
                        <path
                          d="M0 0h12v8H0z"
                        />
                      </g>
                    </g>
                  </svg>
                </ChevronDown>
              </span>
            </div>
            <Menu
              aria-labelledby="downshift-0-label"
              focus={false}
              id="downshift-0-menu"
              isOpen={false}
              role="listbox"
              width={null}
            >
              <ul
                aria-labelledby="downshift-0-label"
                className="emotion-6"
                id="downshift-0-menu"
                role="listbox"
                width={null}
              />
            </Menu>
          </div>
        </Render>
      </SelectOnKeyPressContainer>
    </div>
  </Downshift>
</Dropdown>
`;

exports[`Dropdown renders correctly with props provided 1`] = `
.emotion-10 {
  width: 100%;
  height: 100%;
}

.emotion-5 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  height: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.emotion-11 {
  font-family: Ciutadella;
  font-size: 18px;
  font-weight: 700;
  font-style: normal;
  font-stretch: normal;
  line-height: 1.56;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
  background: #3c3c3c;
  color: #8de2e0;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  position: static;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  height: 20px;
  max-width: 100%;
  outline: none;
}

.emotion-0 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  padding: 0;
}

.emotion-1 {
  border: 0px;
  -webkit-clip: rect(0px,0px,0px,0px);
  clip: rect(0px,0px,0px,0px);
  margin: -1px;
  position: absolute;
  background-color: transparent;
  color: transparent;
  cursor: pointer;
  width: 100%;
  height: 100%;
}

.emotion-1::-webkit-input-placeholder {
  color: inherit;
}

.emotion-1::-moz-placeholder {
  color: inherit;
}

.emotion-1:-ms-input-placeholder {
  color: inherit;
}

.emotion-1::placeholder {
  color: inherit;
}

.emotion-2 {
  width: 24px;
  height: 100%;
  fill: #8de2e0;
  vertical-align: middle;
  padding: 0;
  box-sizing: content-box;
}

.emotion-9 {
  display: none;
  min-width: 240px;
  border-radius: 4px;
  box-shadow: 0 1px 2px 0 rgba(0,0,0,0.2);
  border: solid 1px #dadada;
  background-color: #ffffff;
  list-style: none;
  position: absolute;
  z-index: 5;
  margin: 0;
  top: 0px;
  left: 0px;
  width: 800px;
  box-sizing: border-box;
  padding: 10px;
}

.emotion-6 {
  font-weight: 400;
  background-color: none;
  color: #323232;
  box-sizing: border-box;
  padding: 10px;
  min-height: 50px;
}

.emotion-6:hover,
.emotion-6:active,
.emotion-6:focus {
  background-color: #f4f5f6;
}

<Dropdown
  downShiftProps={
    Object {
      "itemToString": [Function],
    }
  }
  focus={true}
  growMenu={true}
  height="20px"
  highlighted={true}
  inline={true}
  items={
    Array [
      Object {
        "name": "name:apple",
        "value": "value:apple",
      },
      Object {
        "name": "name:orange",
        "value": "name:orange",
      },
      Object {
        "name": "name:banana",
        "value": "name:banana",
      },
    ]
  }
  menuWidth="800px"
  onChange={[Function]}
  renderItem={[Function]}
  withPadding={true}
>
  <Downshift
    defaultHighlightedIndex={null}
    defaultIsOpen={false}
    environment={[Window]}
    getA11yStatusMessage={[Function]}
    initialSelectedItem={
      Object {
        "name": "name:apple",
        "value": "value:apple",
      }
    }
    itemToString={[Function]}
    onChange={[Function]}
    onInputValueChange={[Function]}
    onOuterClick={[Function]}
    onSelect={[Function]}
    onStateChange={[Function]}
    onUserAction={[Function]}
    scrollIntoView={[Function]}
    selectedItemChanged={[Function]}
    stateReducer={[Function]}
    suppressRefError={false}
  >
    <div
      aria-expanded={false}
      aria-haspopup="listbox"
      aria-labelledby="downshift-1-label"
      aria-owns={null}
      className="emotion-11"
      role="combobox"
    >
      <SelectOnKeyPressContainer
        downshiftProps={
          Object {
            "clearItems": [Function],
            "clearSelection": [Function],
            "closeMenu": [Function],
            "getInputProps": [Function],
            "getItemProps": [Function],
            "getLabelProps": [Function],
            "getMenuProps": [Function],
            "getRootProps": [Function],
            "getToggleButtonProps": [Function],
            "highlightedIndex": null,
            "id": "downshift-1",
            "inputValue": "name:apple",
            "isOpen": false,
            "itemToString": [Function],
            "openMenu": [Function],
            "reset": [Function],
            "selectHighlightedItem": [Function],
            "selectItem": [Function],
            "selectItemAtIndex": [Function],
            "selectedItem": Object {
              "name": "name:apple",
              "value": "value:apple",
            },
            "setHighlightedIndex": [Function],
            "setItemCount": [Function],
            "setState": [Function],
            "toggleMenu": [Function],
            "unsetItemCount": [Function],
          }
        }
        focus={false}
        itemToString={[Function]}
        items={
          Array [
            Object {
              "name": "name:apple",
              "value": "value:apple",
            },
            Object {
              "name": "name:orange",
              "value": "name:orange",
            },
            Object {
              "name": "name:banana",
              "value": "name:banana",
            },
          ]
        }
      >
        <Render
          downShiftProps={
            Object {
              "itemToString": [Function],
            }
          }
          downshiftProps={
            Object {
              "clearItems": [Function],
              "clearSelection": [Function],
              "closeMenu": [Function],
              "getInputProps": [Function],
              "getItemProps": [Function],
              "getLabelProps": [Function],
              "getMenuProps": [Function],
              "getRootProps": [Function],
              "getToggleButtonProps": [Function],
              "highlightedIndex": null,
              "id": "downshift-1",
              "inputValue": "name:apple",
              "isOpen": false,
              "itemToString": [Function],
              "openMenu": [Function],
              "reset": [Function],
              "selectHighlightedItem": [Function],
              "selectItem": [Function],
              "selectItemAtIndex": [Function],
              "selectedItem": Object {
                "name": "name:apple",
                "value": "value:apple",
              },
              "setHighlightedIndex": [Function],
              "setItemCount": [Function],
              "setState": [Function],
              "toggleMenu": [Function],
              "unsetItemCount": [Function],
            }
          }
          focus={true}
          growMenu={true}
          height="20px"
          highlighted={true}
          inline={true}
          isFocused={false}
          items={
            Array [
              Object {
                "name": "name:apple",
                "value": "value:apple",
              },
              Object {
                "name": "name:orange",
                "value": "name:orange",
              },
              Object {
                "name": "name:banana",
                "value": "name:banana",
              },
            ]
          }
          label=""
          menuWidth="800px"
          onChange={[Function]}
          placeholder=""
          renderItem={[Function]}
          setFocus={[Function]}
          withPadding={true}
        >
          <div
            className="emotion-10"
          >
            <label
              htmlFor="downshift-1-input"
              id="downshift-1-label"
            />
            <div
              className="emotion-5"
              onClick={[Function]}
            >
              <span
                className="emotion-0"
              >
                name:apple
              </span>
              <input
                aria-activedescendant={null}
                aria-autocomplete="list"
                aria-controls={null}
                aria-labelledby="downshift-1-label"
                autoComplete="off"
                className="emotion-1"
                id="downshift-1-input"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                placeholder=""
                readOnly={true}
                tabIndex="0"
                value="name:apple"
              />
              <span
                className="emotion-2"
              >
                <ChevronDown
                  className="emotion-2"
                  height="24"
                  width="24"
                >
                  <svg
                    className="emotion-2"
                    height="24"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                    xmlnsXlink="http://www.w3.org/1999/xlink"
                  >
                    <defs>
                      <path
                        d="M2.12 1.29L6 5.17l3.88-3.88a.996.996 0 1 1 1.41 1.41L6.7 7.29a.996.996 0 0 1-1.41 0L.7 2.7a.996.996 0 0 1 0-1.41c.39-.38 1.03-.39 1.42 0z"
                        id="mockId"
                      />
                    </defs>
                    <g
                      fillRule="evenodd"
                      transform="translate(6 8)"
                    >
                      <mask
                        id="mockId"
                      >
                        <use
                          xlinkHref="#mockId"
                        />
                      </mask>
                      <use
                        fillRule="nonzero"
                        xlinkHref="#mockId"
                      />
                      <g
                        mask="url(#mockId)"
                      >
                        <path
                          d="M0 0h12v8H0z"
                        />
                      </g>
                    </g>
                  </svg>
                </ChevronDown>
              </span>
            </div>
            <Menu
              aria-labelledby="downshift-1-label"
              focus={false}
              id="downshift-1-menu"
              isOpen={false}
              role="listbox"
              width="800px"
            >
              <ul
                aria-labelledby="downshift-1-label"
                className="emotion-9"
                id="downshift-1-menu"
                role="listbox"
                width="800px"
              >
                <MenuItem
                  aria-selected={false}
                  highlighted={false}
                  id="downshift-1-item-0"
                  key="name:apple-0"
                  onClick={[Function]}
                  onMouseDown={[Function]}
                  onMouseMove={[Function]}
                  role="option"
                  selected={true}
                >
                  <li
                    aria-selected={false}
                    className="emotion-6"
                    id="downshift-1-item-0"
                    onClick={[Function]}
                    onMouseDown={[Function]}
                    onMouseMove={[Function]}
                    role="option"
                    selected={true}
                  >
                    <div>
                      mockRenderItemResult
                    </div>
                  </li>
                </MenuItem>
                <MenuItem
                  aria-selected={false}
                  highlighted={false}
                  id="downshift-1-item-1"
                  key="name:orange-1"
                  onClick={[Function]}
                  onMouseDown={[Function]}
                  onMouseMove={[Function]}
                  role="option"
                  selected={false}
                >
                  <li
                    aria-selected={false}
                    className="emotion-6"
                    id="downshift-1-item-1"
                    onClick={[Function]}
                    onMouseDown={[Function]}
                    onMouseMove={[Function]}
                    role="option"
                    selected={false}
                  >
                    <div>
                      mockRenderItemResult
                    </div>
                  </li>
                </MenuItem>
                <MenuItem
                  aria-selected={false}
                  highlighted={false}
                  id="downshift-1-item-2"
                  key="name:banana-2"
                  onClick={[Function]}
                  onMouseDown={[Function]}
                  onMouseMove={[Function]}
                  role="option"
                  selected={false}
                >
                  <li
                    aria-selected={false}
                    className="emotion-6"
                    id="downshift-1-item-2"
                    onClick={[Function]}
                    onMouseDown={[Function]}
                    onMouseMove={[Function]}
                    role="option"
                    selected={false}
                  >
                    <div>
                      mockRenderItemResult
                    </div>
                  </li>
                </MenuItem>
              </ul>
            </Menu>
          </div>
        </Render>
      </SelectOnKeyPressContainer>
    </div>
  </Downshift>
</Dropdown>
`;
